src/meta/src/rpc/server.rs

```diff
@@ -171,316 +153,312 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
-        tokio::spawn(dashboard_service.serve(address_info.ui_path));
+    /// Check for `number_of_nodes` meta leader nodes, starting at `meta_port`, `meta_port + 1`, ...
+    /// Simulates `number_of_nodes` compute nodes, starting at `meta_port`, `meta_port + 1`, ...
+    ///
+    /// ## Returns
+    /// Number of nodes which currently are leaders. Number is not snapshoted. If there is a
+    /// leader failover in process, you may get an incorrect result
+    async fn _number_of_leaders(number_of_nodes: u16, meta_port: u16, host_port: u16) -> u16 {
+        let mut leader_count = 0;
+        for i in 0..number_of_nodes {
+            let local = "127.0.0.1".to_owned();
+            let port = meta_port + i;
+            let meta_addr = format!("http://{}:{}", local, port);
+            let host_addr = HostAddr {
+                host: local,
+                port: host_port + i,
+            };
+
+            let endpoint = Endpoint::from_shared(meta_addr.to_string())
+                .unwrap()
+                .initial_connection_window_size(MAX_CONNECTION_WINDOW_SIZE);
+            let channel = endpoint
+                .http2_keep_alive_interval(Duration::from_secs(60))
+                .keep_alive_timeout(Duration::from_secs(60))
+                .connect_timeout(Duration::from_secs(5))
+                .connect()
+                .await
+                .inspect_err(|e| {
+                    tracing::warn!(
+                        "Failed to connect to meta server {}, wait for online: {}",
+                        meta_addr,
+                        e
+                    );
+                })
+                .unwrap();
+
+            let cluster_client = ClusterServiceClient::new(channel);
+            let resp = cluster_client
+                .to_owned()
+                .add_worker_node(AddWorkerNodeRequest {
+                    worker_type: WorkerType::ComputeNode as i32,
+                    host: Some(HostAddress {
+                        host: host_addr.host,
+                        port: host_addr.port as i32,
+                    }),
+                    worker_node_parallelism: 5,
+                })
+                .await;
+            let is_leader = resp.is_ok();
+
+            if is_leader {
+                leader_count += 1;
+            }
+        }
+        leader_count
     }
 
-    let catalog_manager = Arc::new(CatalogManager::new(env.clone()).await.unwrap());
+    // Writing these tests as separate functions instead of one loop, because functions get executed
+    // in parallel
+    #[tokio::test]
+    async fn test_single_leader_setup_1() {
+        let v = _setup_n_nodes(1, 1234).await;
+        let leader_count = _number_of_leaders(1, 1234, 5678).await;
+        assert_eq!(
+            leader_count, 1,
+            "Expected to have 1 leader, instead got {} leaders",
+            leader_count
+        );
+        for element in v {
+            element.0.abort();
+        }
```

Comment:
```suggestion
        for (join_handle, shutdown_tx) in v {
            shutdown_tx.send(()).unwrap();
            join_handle.await.unwrap();
        }
```
Can we stop meta in this way?

Reply:
Hey, this was not about stopping the meta nodes. The nodes get stopped automatically once the test is over. The issue was that if we do not pass around the handle, the nodes "go out of scope", because their handle goes out of scope. 

I tried the above approach and the tests do not terminate. I am not sure why. My guess is that something is still connecting against the nodes and that the graceful shutdown prevents test termination.

Reply:
Also, we have to be careful when handling the result of the shutdown signal: Because we already killed a node in the `failover` test, `shutdown_tx.send(())` may result in an error. If we go for above approach, I would prefer 

```rust
let _ = shutdown_tx.send(());
join_handle.await.unwrap();
```

Reply:
Found the issue: Follower nodes were not waiting for shutdown signal, but only for the next election.